services/business/checkers/destination_rules/multi_match_checker.go

```diff
@@ -0,0 +1,44 @@
+package destination_rules
+
+import (
+	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/services/models"
+)
+
+type MultiMatchChecker struct {
+	DestinationRules []kubernetes.IstioObject
+}
+
+func (m MultiMatchChecker) Check() models.IstioValidations {
+	validations := models.IstioValidations{}
+
+	var empty struct{}
+	seenHosts := make(map[string]struct{}, len(m.DestinationRules))
+
+	for _, v := range m.DestinationRules {
+		if host, ok := v.GetSpec()["host"]; ok {
+			if dHost, ok := host.(string); ok {
+				if _, found := seenHosts[dHost]; found {
+					destinationRulesName := v.GetObjectMeta().Name
+					key := models.IstioValidationKey{Name: destinationRulesName, ObjectType: "destinationrules"}
+					checks := models.BuildCheck("More than one DestinationRules for same host",
+						"warning", "spec/hosts")
+					rrValidation := &models.IstioValidation{
+						Name:       destinationRulesName,
+						ObjectType: "destinationrules",
+						Valid:      true,
+						Checks: []*models.IstioCheck{
+							&checks,
+						},
+					}
+
+					validations.MergeValidations(models.IstioValidations{key: rrValidation})
+				} else {
+					seenHosts[dHost] = empty
```

Comment:
I think that using a one-field-key map like this one, you are going to miss some duplicates due to the different ways istio allow users to specify hosts. 
e.g. `reviews` host might appear in a destination rule as `reviews.bookinfo.svc.cluster.local`. 
also `reviews` host can fell into a wildcard host like this one `*.bookinfo.svc.cluster.local' or simply to a `*` host.

Therefore, if you have on DR with `reviews` and another one with `reviews.bookinfo.svc.cluster.local`, this validation won't show up.

We did some work connected to determine duplicated by host for Virtual Services: https://github.com/kiali/kiali/blob/master/services/business/checkers/virtual_services/single_host_checker.go#L122
At the end, we have to compare host not just by name but conceptually (name, namespace, cluster).

Reply:
I tried to search for the rules for those hostnames, but I could not find anything in the documention (now browsing the istio source code). Are all regexps allowed for the destionationrules, or is this the definition:

https://github.com/istio/istio/blob/master/pilot/pkg/model/validation.go#L43

Thus only allowing the leading ``*`` or ``*-`` ?

Reply:
And the secondary question, are these harmful definitions or are they alright:

* *.bookinfo.svc.cluster.local
* bookinfo.svc.cluster.local

It shouldn't match the same hosts, but how does it do sub-hosts internally?

Reply:
I use to follow the descriptions in the reference page: https://istio.io/docs/reference/config/istio.networking.v1alpha3/#DestinationRule

> REQUIRED. The name of a service from the service registry. Service names are looked up from the platform’s service registry (e.g., Kubernetes services, Consul services, etc.) and from the hosts declared by ServiceEntries. Rules defined for services that do not exist in the service registry will be ignored.

> Note for Kubernetes users: When short names are used (e.g. “reviews” instead of “reviews.default.svc.cluster.local”), Istio will interpret the short name based on the namespace of the rule, not the service. A rule in the “default” namespace containing a host “reviews will be interpreted as “reviews.default.svc.cluster.local”, irrespective of the actual namespace associated with the reviews service. To avoid potential misconfigurations, it is recommended to always use fully qualified domain names over short names.

From that description, I understand that  you can have both 'short name' (reviews) and 'FQDN' (reviews.bookinfo.svc.cluster.local). What I can't tell is whether wildcard expressions are allowed or no. It would be great to prepare an example mesh with that an play with it.

Reply:
I don't think that wildcards will work well for DestinationRule.
So I would consider a first scenario where inner service can be expressed in short name or FQDN.
If wildcards are supported at this level, that can be clarified and fixed later.

Reply:
This seems to indicate that wildcard could be usable? https://github.com/istio/istio/blob/master/tests/testdata/config/destination-rule-all.yaml#L24

Reply:
Yeah, it is allowed, from the Istio blog: "The host of a DestinationRule can include wildcard prefixes, allowing a single rule to be specified for many actual services."